File: 5-object-factory/node-addon-api/addon.cc
#include <napi.h>

Napi::Object CreateObject(const Napi::CallbackInfo& info) {
  Napi::Env env = info.Env();

  Napi::Object obj = Napi::Object::New(env);

  if (!info[0].IsString()) {
    Napi::TypeError::New(env, "Wrong argument! Should be a string").ThrowAsJavaScriptException();
    return obj;
  }

  obj.Set(Napi::String::New(env, "msg"), info[0].ToString());

  return obj;
}

Napi::Object Init(Napi::Env env, Napi::Object exports) {
  return Napi::Function::New(env, CreateObject, "createObject");
}

NODE_API_MODULE(addon, Init)

File: 10-factory-wrap/napi/addon.js
let createObject = require('bindings')('addon');

console.log(createObject);

let obj = createObject(10);
console.log(obj.plusOne()); // 11
console.log(obj.plusOne()); // 12
console.log(obj.plusOne()); // 13

let obj2 = createObject(20);
console.log(obj2.plusOne()); // 21
console.log(obj2.plusOne()); // 22
console.log(obj2.plusOne()); // 23

let obj3 = createObject('20'); // Should throw TypeError

